prov/gni/src/gnix_mr.c

```diff
@@ -297,21 +297,33 @@ static inline void *__gnix_generic_register(
 	gni_return_t grc = GNI_RC_SUCCESS;
 	int rc;
 
+	pthread_mutex_lock(&gnix_nic_list_lock);
+
 	/* If the nic list is empty, create a nic */
-        if (unlikely((dlist_empty(&gnix_nic_list_ptag[domain->ptag])))) {
-                rc = gnix_nic_alloc(domain, NULL, &nic);
-                if (rc) {
-                        GNIX_INFO(FI_LOG_MR,
-                                  "could not allocate nic to do mr_reg,"
-                                  " ret=%i\n", rc);
-                        return NULL;
-                }
-        } else {
-                nic = dlist_first_entry(&gnix_nic_list_ptag[domain->ptag], 
+	if (unlikely((dlist_empty(&gnix_nic_list_ptag[domain->ptag])))) {
+		/* release the lock because we are not checking the list after
+			this point. Additionally, gnix_nic_alloc takes the 
+			lock to add the nic. */
+		pthread_mutex_unlock(&gnix_nic_list_lock);
+
+		rc = gnix_nic_alloc(domain, NULL, &nic);
+		if (rc) {
+			GNIX_INFO(FI_LOG_MR,
+				  "could not allocate nic to do mr_reg,"
+				  " ret=%i\n", rc);
+			return NULL;
+		}
+	} else {
+		nic = dlist_first_entry(&gnix_nic_list_ptag[domain->ptag], 
 			struct gnix_nic, ptag_nic_list);
-                if (!nic)
-                        return NULL;
+		if (unlikely(nic == NULL)) {
+			GNIX_INFO(FI_LOG_MR, "Failed to find nic on "
```

Comment:
I'm going to change the INFO to ERROR, and it perhaps should even be a FATAL, but I'll leave that open to discussion.